src/rgw/cls_fifo_legacy.cc

```diff
@@ -1679,6 +1679,7 @@ int FIFO::list(const DoutPrefixProvider *dpp, int max_entries,
 
     std::unique_lock l(m);
     auto part_oid = info.part_oid(part_num);
+    bool at_head = part_num >= info.head_part_num;
```

Comment:
This change will be a NO-OP.
so looking the current race condition issue, where the 2 parts will be created but will not be part of link list, the orphan part will go and create a part and set the head to itself, and the other thread/process will do the same, create part and set the head_part_num to itself.
So at `at_head` will ALWAYS be TRUE for orphan part, because based on condition part_num will be passed by secondary which will be orphan part num and the info.head_part_num will also be the orphan part num, because while creation of new part and setting of head it will set the head to itself. 
 there needs to be way to find out GLOBAL_HEAD or max_head value and if orphan head_part_num >= GLOBAL_HEAD then at_head will be false and listing will continue to go and read the next part

Reply:
What link list? If you think we need to update the metadata, you might be right. I'm considering that we might need to use watch/notify, which I would have preferred to avoid.